### PR TITLE
Improve republish documents

### PR DESCRIPTION
**Pull request by jamiecobbett**
Equivalent to: https://github.com/alphagov/specialist-publisher/pull/483

Hat tip to @heathd for helping with this.

Successfully tested in development.

### CODE REVIEW

File: app/services/abstract_document_service_registry.rb [hunk 1]
```diff
@@ -72,6 +72,20 @@ def publish(document_id)
...
 
+  def republish(document_id)
+    RepublishDocumentService.new(
+      document_repository: document_repository,
+      listeners: observers.republication,
+      document_id: document_id,
+    )
+  end
+
+  def republish_all
```
Comment (evilstreak) [inline app/services/abstract_document_service_registry.rb]: It seems odd to have this method perform the action when every other public method in this class returns a callable.

File: app/services/republish_document_service.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    listeners.each { |l| l.call(document) }
+  end
+
+  def document
+    @document ||= document_repository.fetch(document_id)
+  rescue KeyError => error
+    raise DocumentNotFoundError.new(error)
+  end
+
+  class ManualNotFoundError < StandardError; end
```
Comment (evilstreak) [inline app/services/republish_document_service.rb]: Should this be `DocumentNotFoundError`?

Comment (jamiecobbett) [inline app/services/republish_document_service.rb]: Fixed and rebased.

File: app/services/abstract_document_service_registry.rb (hunk 1 above)
Comment (jamiecobbett) [inline app/services/abstract_document_service_registry.rb]: Good point. We could make it return a callable?
```

Comment (evilstreak) [inline app/services/abstract_document_service_registry.rb]: Sounds good, though I'd probably use a lambda for consistency with the rest of this code base: `-> { … }`

Comment (jamiecobbett) [inline app/services/abstract_document_service_registry.rb]: Fixed and rebased.

**Comment by evilstreak** [comment]
:+1:

**Pull request merged**